### ISSUE TITLE
Is there a way to use CMYK color to draw shapes?

### ISSUE THREAD

**Issue by OleksandrRakovets** [pymupdf/PyMuPDF#270]
First of all, I want to thank the authors for the great library! I am relatively new to Python and PDF manipulation and your lib makes my life easier.

I work in the printing industry hence all PDFs generally use CMYK. I have managed to draw required markup shapes over existing files, but they won't pass printing software validation due to colorspace.

I could not find anything related to colorspace in the docs. Do you have any advice on how to achieve drawing lines with a CMYK color?

**Comment by JorjMcKie** [comment]
Oha!
You are even the first one who ever mentioned to use my drawing methods!
To be honest, I have just been too lazy to implement CMYK.
I do support it with annotation form fields (widgets), but not (yet) for drawing.
To make sure there is no miscommunication: we are talking about things like `Page.drawRect`, etc. right?

To make things simple, I am currently always just using RGB color triples.
As per PDF it is of course possible to use color quadruples - I would have to change the draw methods ... and possibly text insertion, too (?).

Please provide details of what you require, and I can then check what might be the effort and a potential schedule.

What I will probably not be able to implement any time soon is implementation of shading patterns.

**Comment by OleksandrRakovets** [comment]
Yeah, you are right. Here's my current code for markup:

        magenta = (230 / 255, 0, 125 / 255)

        page.drawLine(x1y0, x1yn, magenta, width=0.5)
        page.drawLine(x2y0, x2yn, magenta, width=0.5)
        page.drawLine(x0y1, xny1, magenta, width=0.5)
        page.drawLine(x0y2, xny2, magenta, width=0.5)

As you just described I need to set the color in CMYK quadruple fashion:

       magenta = (0, 100, 0, 0)

**Comment by JorjMcKie** [comment]
And it would be sufficient to allow 4-tuples / 4-lists at every place, where currently ``color=...`` or ``fil=...`` parameters with triples of float occur?

Anything about colors for annotations?

**Comment by OleksandrRakovets** [comment]
> And it would be sufficient to allow 4-tuples / 4-lists at every place, where currently color=... or fil=... parameters with triples of float occur?

Indeed.

> Anything about colors for annotations?

As far as I understand they might as well be assigned color from different color spaces. My task does not require to use CMYK for them :-)

I would be happy to test this change prior to release if necessary.

**Comment by JorjMcKie** [comment]
Would be great if you could help with that  ... and very possible if your system is Windows. Or if you otherwise can generate Python binaries for your system.

**Comment by JorjMcKie** [comment]
If it is Windows, then please do this

```python
D:\Jorj\Desktop>python
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
```
and include it here. It would send you a wheel which you can install with PIP.

**Comment by OleksandrRakovets** [comment]
This is from work windows PC.

```
C:\Users\prinect>python
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 22:20:52) [MSC v.1916 32 bit (Intel)] on win32
```
Usually I run Python code inside Docker containers on my Mac.

**Comment by JorjMcKie** [comment]
Hm, not sure how this can work.
After I am done, I could send you a wheel file ``PyMuPDF-1.14.12-cp37-cp37m-win32.whl`` which needs to be installed like that: ``python -m pip install -U PyMuPDF-1.14.12-cp37-cp37m-win32.whl``.
Then you have a new PyMuPDF version 1.14.12 on your Windows system.

**Comment by JorjMcKie** [comment]
here is a first cut:

[PyMuPDF-1.14.12-cp37-cp37m-win32.whl.zip](https://github.com/pymupdf/PyMuPDF/files/2988399/PyMuPDF-1.14.12-cp37-cp37m-win32.whl.zip)

As discussed, it is now possible to specify different color / fill schemes:
* float or (float,) tuple: gray
* float triple: RGB
* float quadruple: CMYK

All floats must be in range [0, 1] (error messages in this regard may not yet correctly reflect to the colorspace)

**Comment by OleksandrRakovets** [comment]
Perfect! It has worked as expected. Acrobat displayed my color lines as CMYK.

**Comment by JorjMcKie** [comment]
good!
Also did much testing, so I will publish this patch version in the next few days.

Thanks for bringing this up and for your help with testing.

**Comment by JorjMcKie** [comment]
fixed with v1.14.12
